### ISSUE TITLE
Segmentation fault when 2 different tables lock at one time

### ISSUE THREAD

**Issue by krsvital** [efficient/libcuckoo#102]
example:

typedef cuckoohash_map<std::string,std::string> t;
t table1;
t table2;

boost::thread *td = new boost::thread([&](){
    try {
          someFunction();
     }
      catch(boost::exception &e){
                 throw;
         }
 });

td->detach();

auto lt = table1.lock_table();
for(auto const& item : lt){
 //do somthing
}

..........................................

void someFunction(){
  auto lt = table2.lock_table();
    for(auto const& item : lt){
      //do somthing
    }
}

**Comment by manugoyal** [comment]
Hi krsvital, I suspect the issue might be with the detached thread using deleted memory. I wrote two programs to emulate what your example is doing:

Program 1:
```
#include <iostream>
#include <string>
#include <thread>

#include <libcuckoo/cuckoohash_map.hh>

using tbl_t = cuckoohash_map<std::string, std::string>;
tbl_t table1;
tbl_t table2;

void some_function() {
  auto lt2 = table2.lock_table();
  std::cout << "Printing table2 contents\n";
  for (const auto& item : lt2) {
    std::cout << item.first << " " << item.second << "\n";
  }
}

int main() {
  auto t = std::thread([] {
    some_function();
  });

  auto lt1 = table1.lock_table();
  std::cout << "Printing table1 contents\n";
  for (const auto& item : lt1) {
    std::cout << item.first << " " << item.second << "\n";
  }

  t.detach();
}
```

Program 2 (identical except for the last line):
```
#include <iostream>
#include <string>
#include <thread>

#include <libcuckoo/cuckoohash_map.hh>

using tbl_t = cuckoohash_map<std::string, std::string>;
tbl_t table1;
tbl_t table2;

void some_function() {
  auto lt2 = table2.lock_table();
  std::cout << "Printing table2 contents\n";
  for (const auto& item : lt2) {
    std::cout << item.first << " " << item.second << "\n";
  }
}

int main() {
  auto t = std::thread([] {
    some_function();
  });

  auto lt1 = table1.lock_table();
  std::cout << "Printing table1 contents\n";
  for (const auto& item : lt1) {
    std::cout << item.first << " " << item.second << "\n";
  }

  t.join();
}
```

The second program segfaults on my machine, while the first doesn't. It would seem the detached thread will continue running even after the main thread completes (and deletes the memory of the global tables). Here's an example that demonstrates similar behavior:

```
#include <iostream>
#include <string>
#include <thread>

class A {
 public:
  A() {}
  ~A() {
    std::cout << "Destroying A\n";
  }
  
  int counter;
};

A obj;

int main() {
  auto t = std::thread([] {
    obj.counter = 0;
    for (int i = 0; i < 1000000; ++i) {
      std::cout << "thread i = " << i << "\n";
      obj.counter += i;
    }
    std::cout << "counter = " << obj.counter << "\n";
  });

  t.detach();

  for (int i = 0; i < 100; ++i) {
    obj.counter += i;
  }
  std::cout << "main thread counter = " << obj.counter << "\n";
}
```

On my machine, this prints:
```
main thread counter = 4950
Destroying A
thread i = 0
thread i = 0
thread i = 1
```

Hopefully, this resolves the issue.
